File: C++/Breadth-first Search/1162. As Far from Land as Possible.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int maxDistance(vector<vector<int>> &grid) {
        const int dx[]{-1, 0, 0, 1};
        const int dy[]{0, -1, 1, 0};
        queue<pair<int, int>> queue;
        int n = grid.size(), m = grid[0].size(), result = 0;
        for (int i = 0; i < n; ++i)
            for (int j = 0; j < m; ++j)
                if (grid[i][j]) queue.push({i, j});
        if (queue.empty() || queue.size() == n * m) return -1;
        while (!queue.empty()) {
            result++;
            for (int i = queue.size(); i > 0; --i) {
                auto cur = queue.front();
                queue.pop();
                for (int k = 0; k < 4; k++) {
                    int x = cur.first + dx[k];
                    int y = cur.second + dy[k];
                    if (x < 0 || y < 0 || x >= n || y >= m) continue;
                    if (grid[x][y] == 0) {
                        queue.push({x, y});
                        grid[x][y] = 1;
                    }
                }
            }
        }
        return result;
    }
};

File: Python/bfs/1162. As Far from Land as Possible.py
from typing import List
from queue import Queue


class Solution:
    def maxDistance(self, grid: List[List[int]]) -> int:
        n, m = len(grid), len(grid[0])
        dx = [-1, 0, 0, 1]
        dy = [0, -1, 1, 0]
        res = -1
        queue = Queue()
        for i in range(n):
            for j in range(m):
                if grid[i][j] == 1: queue.put([i, j])
        if queue.queue == 0 or queue.qsize() == m * n: return -1

        while not queue.empty():
            res += 1
            for k in range(queue.qsize()):
                cur = queue.get()
                for l in range(4):
                    new_x, new_y = cur[0] + dx[l], cur[1] + dy[l]
                    if new_x < 0 or new_x >= n or new_y < 0 or new_y >= m: continue
                    if grid[new_x][new_y] == 0:
                        queue.put([new_x, new_y])
                        grid[new_x][new_y] = 1
        return res
